File: packages/android/app/src/main/java/com/billyrinks/antimasking/domain/model/MarketplaceModels.kt
package com.billyrinks.antimasking.domain.model

import java.math.BigDecimal
import java.time.Instant

/**
 * Domain model for marketplace listing
 */
data class MarketplaceListing(
    val id: String,
    val title: String,
    val description: String,
    val price: BigDecimal,
    val currency: Currency = Currency.NGN,
    val category: ListingCategory,
    val images: List<String>,
    val thumbnailUrl: String?,
    val sellerId: String,
    val sellerName: String,
    val state: NigerianState,
    val lga: String? = null,
    val isDiasporaFriendly: Boolean = false,
    val isNegotiable: Boolean = false,
    val viewCount: Int = 0,
    val isFavorite: Boolean = false,
    val status: ListingStatus,
    val createdAt: Instant,
    val updatedAt: Instant
)

enum class ListingCategory(val displayName: String, val emoji: String) {
    ELECTRONICS("Electronics", "📱"),
    VEHICLES("Vehicles", "🚗"),
    PROPERTY("Property", "🏠"),
    FASHION("Fashion", "👗"),
    AGRICULTURE("Agriculture", "🌾"),
    SERVICES("Services", "🛠️"),
    FOOD("Food & Beverages", "🍲"),
    HEALTH("Health & Beauty", "💄"),
    FURNITURE("Furniture", "🪑"),
    JOBS("Jobs", "💼"),
    OTHER("Other", "📦")
}

enum class ListingStatus {
    ACTIVE,
    SOLD,
    RESERVED,
    EXPIRED,
    DELETED
}

/**
 * Nigerian states
 */
enum class NigerianState(val displayName: String, val region: GeopoliticalZone) {
    ABIA("Abia", GeopoliticalZone.SOUTH_EAST),
    ADAMAWA("Adamawa", GeopoliticalZone.NORTH_EAST),
    AKWA_IBOM("Akwa Ibom", GeopoliticalZone.SOUTH_SOUTH),
    ANAMBRA("Anambra", GeopoliticalZone.SOUTH_EAST),
    BAUCHI("Bauchi", GeopoliticalZone.NORTH_EAST),
    BAYELSA("Bayelsa", GeopoliticalZone.SOUTH_SOUTH),
    BENUE("Benue", GeopoliticalZone.NORTH_CENTRAL),
    BORNO("Borno", GeopoliticalZone.NORTH_EAST),
    CROSS_RIVER("Cross River", GeopoliticalZone.SOUTH_SOUTH),
    DELTA("Delta", GeopoliticalZone.SOUTH_SOUTH),
    EBONYI("Ebonyi", GeopoliticalZone.SOUTH_EAST),
    EDO("Edo", GeopoliticalZone.SOUTH_SOUTH),
    EKITI("Ekiti", GeopoliticalZone.SOUTH_WEST),
    ENUGU("Enugu", GeopoliticalZone.SOUTH_EAST),
    FCT("Federal Capital Territory", GeopoliticalZone.NORTH_CENTRAL),
    GOMBE("Gombe", GeopoliticalZone.NORTH_EAST),
    IMO("Imo", GeopoliticalZone.SOUTH_EAST),
    JIGAWA("Jigawa", GeopoliticalZone.NORTH_WEST),
    KADUNA("Kaduna", GeopoliticalZone.NORTH_WEST),
    KANO("Kano", GeopoliticalZone.NORTH_WEST),
    KATSINA("Katsina", GeopoliticalZone.NORTH_WEST),
    KEBBI("Kebbi", GeopoliticalZone.NORTH_WEST),
    KOGI("Kogi", GeopoliticalZone.NORTH_CENTRAL),
    KWARA("Kwara", GeopoliticalZone.NORTH_CENTRAL),
    LAGOS("Lagos", GeopoliticalZone.SOUTH_WEST),
    NASARAWA("Nasarawa", GeopoliticalZone.NORTH_CENTRAL),
    NIGER("Niger", GeopoliticalZone.NORTH_CENTRAL),
    OGUN("Ogun", GeopoliticalZone.SOUTH_WEST),
    ONDO("Ondo", GeopoliticalZone.SOUTH_WEST),
    OSUN("Osun", GeopoliticalZone.SOUTH_WEST),
    OYO("Oyo", GeopoliticalZone.SOUTH_WEST),
    PLATEAU("Plateau", GeopoliticalZone.NORTH_CENTRAL),
    RIVERS("Rivers", GeopoliticalZone.SOUTH_SOUTH),
    SOKOTO("Sokoto", GeopoliticalZone.NORTH_WEST),
    TARABA("Taraba", GeopoliticalZone.NORTH_EAST),
    YOBE("Yobe", GeopoliticalZone.NORTH_EAST),
    ZAMFARA("Zamfara", GeopoliticalZone.NORTH_WEST)
}

enum class GeopoliticalZone(val displayName: String) {
    NORTH_CENTRAL("North Central"),
    NORTH_EAST("North East"),
    NORTH_WEST("North West"),
    SOUTH_EAST("South East"),
    SOUTH_SOUTH("South South"),
    SOUTH_WEST("South West")
}

File: packages/ios/Sources/Domain/Models/MarketplaceModels.swift
import Foundation
import Tagged

// MARK: - Marketplace Listing

public typealias ListingID = Tagged<MarketplaceListing, UUID>

public struct MarketplaceListing: Equatable, Identifiable, Sendable {
    public let id: ListingID
    public let title: String
    public let description: String
    public let price: Decimal
    public let currency: Currency
    public let category: ListingCategory
    public let images: [URL]
    public let thumbnailURL: URL?
    public let sellerId: String
    public let sellerName: String
    public let state: NigerianState
    public let lga: String?
    public let isDiasporaFriendly: Bool
    public let isNegotiable: Bool
    public let viewCount: Int
    public var isFavorite: Bool
    public let status: ListingStatus
    public let createdAt: Date
    public let updatedAt: Date
    
    public init(
        id: ListingID,
        title: String,
        description: String,
        price: Decimal,
        currency: Currency = .ngn,
        category: ListingCategory,
        images: [URL] = [],
        thumbnailURL: URL? = nil,
        sellerId: String,
        sellerName: String,
        state: NigerianState,
        lga: String? = nil,
        isDiasporaFriendly: Bool = false,
        isNegotiable: Bool = false,
        viewCount: Int = 0,
        isFavorite: Bool = false,
        status: ListingStatus = .active,
        createdAt: Date = Date(),
        updatedAt: Date = Date()
    ) {
        self.id = id
        self.title = title
        self.description = description
        self.price = price
        self.currency = currency
        self.category = category
        self.images = images
        self.thumbnailURL = thumbnailURL
        self.sellerId = sellerId
        self.sellerName = sellerName
        self.state = state
        self.lga = lga
        self.isDiasporaFriendly = isDiasporaFriendly
        self.isNegotiable = isNegotiable
        self.viewCount = viewCount
        self.isFavorite = isFavorite
        self.status = status
        self.createdAt = createdAt
        self.updatedAt = updatedAt
    }
}

public enum ListingCategory: String, Equatable, Sendable, CaseIterable {
    case electronics = "ELECTRONICS"
    case vehicles = "VEHICLES"
    case property = "PROPERTY"
    case fashion = "FASHION"
    case agriculture = "AGRICULTURE"
    case services = "SERVICES"
    case food = "FOOD"
    case health = "HEALTH"
    case furniture = "FURNITURE"
    case jobs = "JOBS"
    case other = "OTHER"
    
    public var displayName: String {
        switch self {
        case .electronics: return "Electronics"
        case .vehicles: return "Vehicles"
        case .property: return "Property"
        case .fashion: return "Fashion"
        case .agriculture: return "Agriculture"
        case .services: return "Services"
        case .food: return "Food & Beverages"
        case .health: return "Health & Beauty"
        case .furniture: return "Furniture"
        case .jobs: return "Jobs"
        case .other: return "Other"
        }
    }
    
    public var emoji: String {
        switch self {
        case .electronics: return "📱"
        case .vehicles: return "🚗"
        case .property: return "🏠"
        case .fashion: return "👗"
        case .agriculture: return "🌾"
        case .services: return "🛠️"
        case .food: return "🍲"
        case .health: return "💄"
        case .furniture: return "🪑"
        case .jobs: return "💼"
        case .other: return "📦"
        }
    }
    
    public var iconName: String {
        switch self {
        case .electronics: return "iphone"
        case .vehicles: return "car.fill"
        case .property: return "house.fill"
        case .fashion: return "tshirt.fill"
        case .agriculture: return "leaf.fill"
        case .services: return "wrench.fill"
        case .food: return "fork.knife"
        case .health: return "heart.fill"
        case .furniture: return "sofa.fill"
        case .jobs: return "briefcase.fill"
        case .other: return "shippingbox.fill"
        }
    }
}

public enum ListingStatus: String, Equatable, Sendable, CaseIterable {
    case active = "ACTIVE"
    case sold = "SOLD"
    case reserved = "RESERVED"
    case expired = "EXPIRED"
    case deleted = "DELETED"
}
